tower-buffer/src/worker.rs

```diff
@@ -0,0 +1,215 @@
+use error::ServiceError;
+use futures::future::Executor;
+use futures::{Async, Future, Poll, Stream};
+use message::Message;
+use std::sync::{Arc, Mutex};
+use tokio_sync::mpsc;
+use tower_service::Service;
+
+/// Task that handles processing the buffer. This type should not be used
+/// directly, instead `Buffer` requires an `Executor` that can accept this task.
+pub struct Worker<T, Request>
+where
+    T: Service<Request>,
+{
+    current_message: Option<Message<Request, T::Future, T::Error>>,
+    rx: mpsc::Receiver<Message<Request, T::Future, T::Error>>,
+    service: T,
+    finish: bool,
+    failed: Option<Arc<ServiceError<T::Error>>>,
+    handle: Handle<T::Error>,
+}
+
+/// Get the error out
+pub(crate) struct Handle<E> {
+    inner: Arc<Mutex<Option<Arc<ServiceError<E>>>>>,
+}
+
+/// This trait allows you to use either Tokio's threaded runtime's executor or the `current_thread`
+/// runtime's executor depending on if `T` is `Send` or `!Send`.
+pub trait WorkerExecutor<T, Request>: Executor<Worker<T, Request>>
+where
+    T: Service<Request>,
+{
+}
+
+impl<T, Request, E: Executor<Worker<T, Request>>> WorkerExecutor<T, Request> for E where
+    T: Service<Request>
+{
+}
+
+impl<T, Request> Worker<T, Request>
+where
+    T: Service<Request>,
+{
+    pub(crate) fn spawn<E>(
+        service: T,
+        rx: mpsc::Receiver<Message<Request, T::Future, T::Error>>,
+        executor: &E,
+    ) -> Result<Handle<T::Error>, T>
```

Comment:
This is changed to initialize the handle & return it instead of having a handle passed in.